### PR TITLE
proxy support

### PR DESCRIPTION
**Pull request by arjgupta**

### CODE REVIEW

File: Tasks/Common/azurerest-common/azurerestutility.ts [hunk 1]
```diff
@@ -348,14 +363,20 @@ async function getOperationStatus(SPN, url: string) {
...
-            deferred.resolve({ "response": response, "content": body } );
-        }
+
+    let options: rm.IRequestOptions = {};
+    options.additionalHeaders = headers;
+    let promise: Promise<rm.IRestResponse<any>> = rc.get(url, options);
+    promise.then((response) => {
+        deferred.resolve({
+            "response": response.statusCode,
+            "content": response.result
```
Comment (dikhakha) [inline Tasks/Common/azurerest-common/azurerestutility.ts]: Will this include Error if present in response body ?

File: Tasks/Common/azurerest-common/azurerestutility.ts [hunk 2]
```diff
@@ -348,14 +363,20 @@ async function getOperationStatus(SPN, url: string) {
...
-        else {
-            deferred.resolve({ "response": response, "content": body } );
-        }
+
+    let options: rm.IRequestOptions = {};
+    options.additionalHeaders = headers;
+    let promise: Promise<rm.IRestResponse<any>> = rc.get(url, options);
+    promise.then((response) => {
+        deferred.resolve({
+            "response": response.statusCode,
```
Comment (dikhakha) [inline Tasks/Common/azurerest-common/azurerestutility.ts]: Can you make corresponding changes in the monitorSlotSwap method to use 'response' directly instead of 'response.statusCode' and 'response.StatusMessage'

File: Tasks/AzureRmWebAppDeployment/kuduutility.ts [hunk 3]
```diff
@@ -297,25 +332,27 @@ async function getFileContent(publishingProfile, physicalPath, fileName) {
...
-    httpObj.get('GET', kuduGetFileUrl, headers, (error, response, body) => {
-        if(error) {
-            defer.reject({ error: error + ''});
-        }
-        else if(response.statusCode === 200) {
+    let options: rm.IRequestOptions = {};
+    options.additionalHeaders = headers;
+    let promise: Promise<any> = hc.get(kuduGetFileUrl, options);
+    promise.then(async (response) => {
+        let contents: string = await response.readBody();
```
Comment (vincentdass) [inline Tasks/AzureRmWebAppDeployment/kuduutility.ts]: This call should be under a try-catch. If not then any exception thrown wont be notified.

File: Tasks/Common/azurerest-common/azurerestutility.ts [hunk 4]
```diff
@@ -169,21 +189,24 @@ function getAuthorizationToken(endPoint): Q.Promise<string> {
...
-            deferred.reject(error);
-        }
-        else if (response.statusCode == 200) {
-            deferred.resolve(JSON.parse(body).access_token);
-        }
-        else {
-            deferred.reject(tl.loc('CouldnotfetchaccesstokenforAzureStatusCode', response.statusCode, response.statusMessage));
+    hc.post(authorityUrl, requestData, requestHeader).then(async (response: hm.HttpClientResponse) => {
+        if (response.message.statusCode === 200) {
+            let contents: string = await response.readBody();
```
Comment (vincentdass) [inline Tasks/Common/azurerest-common/azurerestutility.ts]: under try-catch block

File: Tasks/AzureRmWebAppDeployment/kuduutility.ts [hunk 5]
```diff
@@ -2,8 +2,23 @@ import Q = require('q');
...
+var requestOptions: httpInterfaces.IRequestOptions = proxyUrl ? { 
+    proxy: { 
+        proxyUrl: proxyUrl, 
+        proxyUsername: tl.getVariable("agent.proxyusername"), 
+        proxyPassword: tl.getVariable("agent.proxypassword"), 
+        proxyBypassHosts: tl.getVariable("agent.proxybypasslist") ? JSON.parse(tl.getVariable("agent.proxybypasslist")) : null 
+    } 
+} : null; 
+
+let hc = new hm.HttpClient(tl.getVariable("AZURE_HTTP_USER_AGENT"), null, requestOptions);
```
Comment (vincentdass) [inline Tasks/AzureRmWebAppDeployment/kuduutility.ts]: just a suggestion, Can we have some name like htpClient?

File: Tasks/AzureResourceGroupDeployment/operations/ResourceGroup.ts [hunk 6]
```diff
@@ -12,10 +12,21 @@ import { PowerShellParameters, NameValuePair } from "./ParameterParser";
 import utils = require("./Utils");
 import fileEncoding = require('./FileEncoding');
 import { ParametersFileObject, TemplateObject, ParameterValue } from "../models/Types";
-
+var hm = require("typed-rest-client/HttpClient");
```
Comment (ds-ms) [inline Tasks/AzureResourceGroupDeployment/operations/ResourceGroup.ts]: Sort imports, keep all imports above and var below

File: Tasks/Common/webdeployment-common/package.json [hunk 7]
```diff
@@ -15,9 +15,8 @@
   "dependencies": {
     "archiver": "1.2.0",
     "decompress-zip": "0.3.0",
-    "ltx": "2.6.2",
+    "ltx": "2.5.0",
```
Comment (vincentdass) [inline Tasks/Common/webdeployment-common/package.json]: why is this changed?

File: Tasks/AzureRmWebAppDeployment/kuduutility.ts [hunk 8]
```diff
@@ -297,25 +332,25 @@ async function getFileContent(publishingProfile, physicalPath, fileName) {
...
     var kuduGetFileUrl = "https://" + publishingProfile.publishUrl + "/api/vfs/" + physicalPath + "/" + fileName;
     tl.debug('Getting content of file: ' + fileName + ' using publishUrl: ' + kuduGetFileUrl);
-    httpObj.get('GET', kuduGetFileUrl, headers, (error, response, body) => {
-        if(error) {
-            defer.reject({ error: error + ''});
-        }
-        else if(response.statusCode === 200) {
+    let promise: Promise<any> = hc.get(kuduGetFileUrl, headers);
+    promise.then(async (response) => {
+        let contents: string = await response.readBody();
```
Comment (vincentdass) [inline Tasks/AzureRmWebAppDeployment/kuduutility.ts]: add try-catch here

File: Tasks/AzureResourceGroupDeployment/operations/ResourceGroup.ts [hunk 9]
```diff
@@ -265,16 +276,16 @@ export class ResourceGroup {
...
     private downloadFile(url): Promise<string> {
         return new Promise<string>((resolve, reject) => {
-            httpObj.get("GET", url, {}, (error, result, contents) => {
-                if (error) {
-                    return reject(tl.loc("FileFetchFailed", url, error));
-                }
-                if (result.statusCode === 200)
+            hc.get(url, {}).then(async (response) => {
+                if(response.message.statusCode == 200) {
+                    let contents: string = await response.readBody();
```
Comment (vincentdass) [inline Tasks/AzureResourceGroupDeployment/operations/ResourceGroup.ts]: try-catch here too

File: Tasks/Common/azurerest-common/azurerestutility.ts [hunk 10]
```diff
@@ -377,24 +400,23 @@ export async function updateAzureRMWebAppConfigDetails(endPoint, webAppName: str
...
-		}
-	});
+    let options: rm.IRequestOptions = {};
+    options.additionalHeaders = headers;
+    let promise: Promise<any> = rc.update(configUrl, JSON.parse(configDetails), options);
+    promise.then((response) => {
+        if(response.statusCode === 200) {
+            deferred.resolve();
+        } else {
+            let errorMessage = "Status Code : " + response.statusCode + "\n";
```
Comment (vincentdass) [inline Tasks/Common/azurerest-common/azurerestutility.ts]: Can you please loc this?
I know this is not your change, can u fix it?

File: Tasks/Common/webdeployment-common/package.json (hunk 7 above)
Comment (arjgupta) [reply Tasks/Common/webdeployment-common/package.json]: current build of the task from master bundles ltx 2.5.0, and not 2.6.2 . Not sure why it was there in the first place.

Comment (arjgupta) [reply Tasks/Common/webdeployment-common/package.json]: found the reason, 2.5.0 is the version that is mentioned in shrinkwrap. whereas 2.6.2 is the one mentioned in package.json. and shrinkwrap takes precedence
